File: src-tauri/src/services/logger.rs
use tauri_plugin_log::{Target, TargetKind, TimezoneStrategy};

/// Builds the tauri-plugin-log plugin with unified format and targets.
///
/// Log format: `[YYYY-MM-DD][HH:MM:SS][target][LEVEL] message`
/// - Backend modules appear as their Rust module path (e.g. `termix::services::sftp_manager`)
/// - Frontend logs from `logger.ts` appear with target `webview` and `[module]` prefix in message
///
/// Targets: Stdout, LogDir (persistent file), Webview (DevTools console)
pub fn build_log_plugin() -> tauri::plugin::TauriPlugin<tauri::Wry> {
    tauri_plugin_log::Builder::new()
        .timezone_strategy(TimezoneStrategy::UseLocal)
        .level(log::LevelFilter::Info)
        .targets([
            Target::new(TargetKind::Stdout),
            Target::new(TargetKind::LogDir { file_name: None }),
            Target::new(TargetKind::Webview),
        ])
        .max_file_size(50_000)
        .build()
}

File: packages/client/src/lib/logger.ts
/**
 * Centralized frontend logger — mirrors the backend logger.rs configuration.
 *
 * All logs flow through `@tauri-apps/plugin-log` to the unified Tauri logging pipeline:
 *   - Stdout (terminal)
 *   - LogDir (persistent file)
 *   - Webview (browser DevTools console)
 *
 * Backend format:  [YYYY-MM-DD][HH:MM:SS][rust_module][LEVEL] message
 * Frontend format: [YYYY-MM-DD][HH:MM:SS][webview][LEVEL] [module] message
 *
 * Usage:
 *   import { createLogger } from "@/lib/logger";
 *   const logger = createLogger("sftp");
 *   logger.info("Connected to", host);
 *   logger.error("Transfer failed:", err);
 */
import {
    debug as tauriDebug,
    error as tauriError,
    info as tauriInfo,
    trace as tauriTrace,
    warn as tauriWarn,
} from "@tauri-apps/plugin-log";

type LogFn = (message: string, ...args: unknown[]) => void;

export interface Logger {
  error: LogFn;
  warn: LogFn;
  info: LogFn;
  debug: LogFn;
  trace: LogFn;
}

function stringify(a: unknown): string {
  if (typeof a === "string") return a;
  if (a instanceof Error) return a.message;
  try {
    return JSON.stringify(a);
  } catch {
    return String(a);
  }
}

function formatMessage(module: string, message: string, args: unknown[]): string {
  const suffix = args.length > 0 ? " " + args.map(stringify).join(" ") : "";
  return `[${module}] ${message}${suffix}`;
}

/**
 * Creates a scoped logger with the given module name.
 * All log calls will be prefixed with `[module]`.
 */
export function createLogger(module: string): Logger {
  return {
    error(message, ...args) {
      tauriError(formatMessage(module, message, args));
    },
    warn(message, ...args) {
      tauriWarn(formatMessage(module, message, args));
    },
    info(message, ...args) {
      tauriInfo(formatMessage(module, message, args));
    },
    debug(message, ...args) {
      tauriDebug(formatMessage(module, message, args));
    },
    trace(message, ...args) {
      tauriTrace(formatMessage(module, message, args));
    },
  };
}

/** Default app-level logger. */
export const log = createLogger("app");
